pkgs/top-level/unit-files.nix

```diff
@@ -0,0 +1,89 @@
+/*
+Internal Nixpkgs file, it may be removed or changed at any point without warning
+
+This file contains the mapping from attribute name to the file in pkgs/unit that contains a definition of it.
+No validity checks are performed, instead this file is optimized for performance.
+TODO: Link to manual and validity checks
+
+Example: If `pkgs/unit/he/hello` exists, the result will have a `hello = pkgs/unit/he/hello/package.nix` entry.
+
+Type: AttrsOf Path
+*/
+
+# Because the result might be fairly large, we only want compute it once.
+# In order to ensure this, we make use of Nix's `import` cache.
+# This cache memoizes the resulting value of the imported file, notably including any `let in` bindings at the top of the file.
+# This is why everything here is computed in a `let in` binding at the top.
+let
+
+  # We don't _need_ `lib` here, but importing it wouldn't be detrimental to performance either because of Nix's `import` cache
+  inherit (builtins) attrNames elemAt length mapAttrs readDir;
+
+  # Warning: Do not make this configurable, otherwise the `import` evaluation cache won't work!
+  baseDirectory = ../unit;
+
+  entries = readDir baseDirectory;
+  shards = attrNames entries;
+
+  # Return the name mappings in a single shard
+  # Type: String -> AttrsOf Path
+  namesForShard = shard:
```

Comment:
If this function returned a list of name value pairs instead, and calling `builtins.listToAttrs` only after merging all lists, what would be the performance characteristic of that? Would it allow us to get rid of the binaryMerge? Especially given that `builtins.concatLists` exists, but maybe I'm naive to hope that it would be faster because of that.

Reply:
Lists unfortunately have the same performance characteristics as attribute sets: `++` has complexity $O(l + r)$ with $l$ and $r$ being the sizes of the left and right side respectively. So because you'd still need to `foldl'` the lists using `++`, it would give poorer performance again.

Reply:
Oh right, that might work! I'll give it a go when I work on this the next time!